### ISSUE TITLE
OAuth token not resent after temporary redirect

### ISSUE THREAD

**Issue by coadler** [github/hub#1859]
After Hub follows a temporary redirect, it does not seem to set the `Authorization` header for the subsequent request

I'm using GitHub enterprise.
```
HUB_VERBOSE=1 hub pull-request                                             
$ git rev-parse -q --git-dir
$ git remote -v
$ git config --get-all hub.host
> GET https://git.foo.com/api/v3/user
> Authorization: token [REDACTED]
> Accept: application/vnd.github.v3+json;charset=utf-8
< HTTP 307
< Location: https://gh.foo.com/api/v3/user
<a href="https://gh.foo.com/api/v3/user">Temporary Redirect</a>.


> GET https://gh.foo.com/api/v3/user
> Accept: application/vnd.github.v3+json;charset=utf-8
< HTTP 401
{"message":"Must authenticate to access this API.","documentation_url":"https://developer.github.com/enterprise/2.14/v3"}
```

**Comment by mislav** [comment]
Hi thank for your reporting! Omitting authorization info on redirect prevents it accidentally leaking to a 3rd-party. In this case, it looks like it's different subdomains under `foo.com` for your org? Would an acceptable solution be forwarding authorization info within the same `foo.com` domain and all its subdomains, but not to any other domains?

Just wondering how your GHE is set up. Thank you!

**Comment by coadler** [comment]
Yep, that would work great. I wasn't thinking about the security implications of this but it definitely makes sense the way you guys did it.

If you're looking for someone to contrib this change I'll happily implement it

**Comment by mislav** [comment]
You're welcome to take a stab. There are some tests under `features/` that exercise different redirect scenarios, but there is nothing like redirect with authorization yet.

Also, I can't remember whether we handle 30x redirects from our code or do we reply on built-in handling of `het/http` standard library. 🤔

**Comment by mislav** [comment]
@coadler We used to manually handle redirects but since then we'd given that up in favor of Go's built-in handling 71fa8dad8faef67599a744df969375ada9694e2c. The `Authorization` header is not forwarded by design.

If we were to forward it, should we allow forwarding the token from `git.foo.com` to any domain under `*.foo.com`? How would we extract the `.foo.com` part? We could perhaps use something like https://github.com/joeguo/tldextract. But now I'm begining to think that all this is more trouble than it's worth.

Does your Enterprise setup still work the same way re: redirects?

**Comment by coadler** [comment]
I was able to work around the problem by just changing the remote URL in `.git` to the domain it was being redirected to. I think this is probably the better fix anyways. I began implementing the forwarding a while ago but unfortunately lost pretty much all of my free time and never got around to finishing it.

Since the error is very confusing at first, do you think it would be feasible to add a warning when hub follows a redirect? I'm not sure if there is a scenario which would absolutely require redirects to work

**Comment by mislav** [comment]
Yes I think hub's error messaging should be improved here to let you know of the redirect and have you update your repository's configuration. Thank you for the info!
